design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,329 @@
+incompatible way (code built using the old `Udaf` interface will not compile, but migration is
+straightforward):
+```java
+/**
+* ...
+* @param <V> value type (the value that is accepted)
+* @param <A> aggregate type
+* @param <VR> reduce type (the value to return)
+*/
+public interface Udaf<V, A, VR> {
+  
+  ...
+ 
+  /**
+   * Redcues an aggregate into the return type.
+   * 
+   * @param aggregate the running aggregate
+   * @return the final return value
+   */ 
+  VR reduce(A aggregate);
+  
+}
+```
+
+`reduce` will be taken as the behavioral parameter to a `mapValues` task that will be applied after
+the aggregate task in the generated KStreams topology.
+
+## Test plan
+
+Each of these improvements will come with some built-in UDF/UDAF examples. These examples will then
+be comprehensively covered by both unit tests as well as corresponding Query Translation Tests.
+Specifically, the UDFs described in the [Value](#value) section will all be implemented alongside
+this KLIP. For the new UDAF complex aggregation, we will ensure that the generated topologies do not
+include any unexpected steps such as a repartition.
+
+## Documentation Updates
+
+### Varargs
+
+* The `Example UDF Class` in `udf.rst` will be extended to showcase the new features. Namely, we
+will replace the multiply double method to include varargs:
+
+>```java
+>      @Udf(description = "multiply N non-nullable DOUBLEs.")
+>      public double multiply(final double... values) {
+>        return Arrays.stream(values).reduce((a, b) -> a * b);;
+>      }
+>```
+
+* The `DESCRIBE FUNCTION` for vararg functions will need to be updated to use the `...` syntax:
+> ```
+> Name        : ARRAYCONTAINS
+> Author      : Confluent
+> Type        : scalar
+> Jar         : internal
+> Variations  :
+> 
+> 	Variation   : ARRAYCONTAINS(VARCHAR... )
+> 	Returns     : BOOLEAN
+> ```
+
+### Struct Support for UDF
+
+* The `Supported Types` section in `udf.rst` will include `Struct` and have an updated note at the 
+bottom which reads:
+
+> Note: Complex types other than List, Map and Struct are not currently supported. Generic types
+> for List and Map are supported, but component elements must be of one of the supported types. If
+> the types are generic, the output type must be able to be inferred from one or more of the input
+> parameters.
+
+* Function descriptions as part of `DESCRIBE FUNCTION` should also show the schema for any `Struct`
+that is used:
+
+> ```
+> Name        : FOO
+> Author      : Confluent
+> Type        : scalar
+> Jar         : internal
+> Variations  :
+>
+>       Variation   : FOO(STRUCT<'VAL' BIGINT, 'OTHER' VARCHAR>)
+>       Returns     : STRUCT<'VAL' BIGINT, 'ANOTHER' LIST<VARCHAR>)
+> ```
+
+### Generics
+
+* The `DESCRIBE FUNCTION` should indicate when a function accepts generics using `ANY<?>` where `?`
+is a letter that begins with `T` and is incremented for each inferred type.
```

Comment:
What happens if we run out of letters? Does it perhaps make sense to use T1, T2, ... instead?

Reply:
That's a better idea!

Reply:
Looks like the doc needs to incorporate that suggestion!